librz/core/cbin.c

```diff
@@ -2305,6 +2305,748 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+			RzStrBuf *buf = rz_strbuf_new(rz_str_get(reloc_name));
+			free(reloc_name);
+			RZ_FREE(name);
+			if (reloc->addend) {
+				if ((reloc->import || reloc->symbol) && !rz_strbuf_is_empty(buf) && reloc->addend > 0) {
+					rz_strbuf_append(buf, " +");
+				}
+				if (reloc->addend < 0) {
+					rz_strbuf_appendf(buf, " - 0x%08" PFMT64x, -reloc->addend);
+				} else {
+					rz_strbuf_appendf(buf, " 0x%08" PFMT64x, reloc->addend);
+				}
+			}
+			if (reloc->is_ifunc) {
+				rz_strbuf_append(buf, " (ifunc)");
+			}
+			char *res = rz_strbuf_drain(buf);
+			if (have_targets) {
+				rz_table_add_rowf(state->d.t, "XXXss", addr, reloc->paddr, reloc->target_vaddr,
+					bin_reloc_type_name(reloc), res);
+			} else {
+				rz_table_add_rowf(state->d.t, "XXss", addr, reloc->paddr,
+					bin_reloc_type_name(reloc), res);
+			}
+			free(res);
+			break;
+		default:
+			rz_warn_if_reached();
+			break;
+		}
+	}
+	rz_cmd_state_output_array_end(state);
+}
+
+static ut64 get_section_addr(RzCore *core, RzBinObject *o, RzBinSection *section) {
+	int va = (core->io->va || core->bin->is_debugger) ? VA_TRUE : VA_FALSE;
+	if (va && !(section->perm & RZ_PERM_R)) {
+		va = VA_NOREBASE;
+	}
+	return rva(o, section->paddr, section->vaddr, va);
+}
+
+static void sections_print_json(RzCore *core, PJ *pj, RzBinObject *o, RzBinSection *section, RzList *hashes) {
+	ut64 addr = get_section_addr(core, o, section);
+	char perms[5];
+	section_perms_str(perms, section->perm);
+
+	pj_o(pj);
+	pj_ks(pj, "name", section->name);
+	pj_kN(pj, "size", section->size);
+	pj_kN(pj, "vsize", section->vsize);
+	pj_ks(pj, "perm", perms);
+	if (!section->is_segment) {
+		char *section_type = rz_bin_section_type_to_string(core->bin, section->type);
+		if (section_type) {
+			pj_ks(pj, "type", section_type);
+		}
+		free(section_type);
+	}
+	if (!section->is_segment) {
+		RzList *flags = rz_bin_section_flag_to_list(core->bin, section->flags);
+		if (!rz_list_empty(flags)) {
+			RzListIter *it;
+			char *pos;
+			pj_ka(pj, "flags");
+			rz_list_foreach (flags, it, pos) {
+				pj_s(pj, pos);
+			}
+			pj_end(pj);
+		}
+		rz_list_free(flags);
+	}
+	pj_kN(pj, "paddr", section->paddr);
+	pj_kN(pj, "vaddr", addr);
+	if (section->align) {
+		pj_kN(pj, "align", section->align);
+	}
+	if (hashes && section->size > 0) {
+		ut8 *data = malloc(section->size);
+		if (data) {
+			ut32 datalen = section->size;
+			RzListIter *iter;
+			char *hashname;
+
+			rz_io_pread_at(core->io, section->paddr, data, datalen);
+
+			rz_list_foreach (hashes, iter, hashname) {
+				char *chkstr = rz_msg_digest_calculate_small_block_string(hashname, data, datalen, NULL, false);
+				if (!chkstr) {
+					continue;
+				}
+				pj_ks(pj, hashname, chkstr);
+				free(chkstr);
+			}
+			free(data);
+		}
+	}
+	pj_end(pj);
+}
+
+static void sections_print_table(RzCore *core, RzTable *t, RzBinObject *o, RzBinSection *section, RzList *hashes) {
+	ut64 addr = get_section_addr(core, o, section);
```

Comment:
Maybe a quick exit if `hashes` is empty?

Reply:
no if there are no hashes we just skip them, but you still print sections/segments.